agent/taskresource/credentialspec/credentialspec.go

```diff
@@ -0,0 +1,556 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package credentialspec
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	apicontainer "github.com/aws/amazon-ecs-agent/agent/api/container"
+	"github.com/aws/amazon-ecs-agent/agent/api/task/status"
+	"github.com/aws/amazon-ecs-agent/agent/credentials"
+	"github.com/aws/amazon-ecs-agent/agent/s3"
+	s3factory "github.com/aws/amazon-ecs-agent/agent/s3/factory"
+	"github.com/aws/amazon-ecs-agent/agent/ssm"
+	ssmfactory "github.com/aws/amazon-ecs-agent/agent/ssm/factory"
+	"github.com/aws/amazon-ecs-agent/agent/taskresource"
+	resourcestatus "github.com/aws/amazon-ecs-agent/agent/taskresource/status"
+	"github.com/aws/amazon-ecs-agent/agent/utils/ioutilwrapper"
+	"github.com/aws/amazon-ecs-agent/agent/utils/oswrapper"
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/cihub/seelog"
+	"github.com/pkg/errors"
+)
+
+const (
+	// ResourceName is the name of the credentialspec resource
+	ResourceName = "credentialspec"
+
+	tempFileName = "temp_file"
+
+	// filePerm is the permission for the credentialspec file.
+	filePerm = 0755
+
+	s3DownloadTimeout = 30 * time.Second
+)
+
+var (
+	// resourceDir will be populated by environment variables
+	resourceDir = ""
+)
+
+func init() {
+	// TODO: Use registry
+	// This should always be available on Windows instances
+	appDataDir := os.Getenv("APPDATA")
+	if appDataDir != "" {
+		resourceDir = appDataDir
+	} else {
+		tempDir := os.Getenv("TEMP")
+		if tempDir != "" {
+			resourceDir = tempDir
+		}
+	}
+
+	if resourceDir == "" {
+		seelog.Critical("Unable to obtain valid credentialspec resource dir")
+		os.Exit(1)
```

Comment:
This will directly exit agent, which we don't want. It should just fail the task. You should be able to just move the logic in **init()** to **NewCredentialSpecResource()**, and the env var should be passed into the function.